README
Joose is an meta object protocol implementation for JavaScript inspired by Moose: http://www.iinteractive.com/moose/

For installation instructions see: INSTALL

For documentation see: doc/html/Joose.html

To run the test suite, open the following url in the browser: t/index.html


Changes
Release history for Joose:

{{ $NEXT }}

    [ENHANCEMENTS]
    
    - none
    
    [API CHANGES]

    - none
    
    [BUG FIXES]
    
    - use feature testing for additional enumeration in Joose.O.each (fixes classes declaration in IE9-Quirks)


3.50.0    2011-09-14 13:51

    [ENHANCEMENTS]
    
    - detect when a class (a function like: `func.meta.c == func`) has been provided as `init` value for the attribute
      and do not call it during initialization (use as "value")
    
    [API CHANGES]

    - removed the undocumented namespaces facility (not documented, low-level stuff)
    
    [BUG FIXES]
    
    - fixed 1 leaked global
    - fixed a bug, preventing roles to have `toString` (and other non-enumerable) methods 


3.018     2011-03-12 12:58


    [ENHANCEMENTS]
    
    - added a sanity checks before consuming a role or inheriting from superclass (should evalutes to "true")
    - from now on, include the "Task/Joose/Core" bundle in the root of distribution as "task-joose-core.js"
    - added a `Joose.C` symbol for configuration options; can be specified as the JOOSE_CFG prior Joose has been loaded
    
    [API CHANGES]

    - removed support for class extension via `Class` keyword - now it will throw the "Double declaration of [ClassName]" exception
      use ClassName.meta.extend() instead
    
    [BUG FIXES]
    
    - updated copyright years in docs


3.017     2011-01-12 18:33

    [ENHANCEMENTS]
    
    - added `strToClass` helper function
    - added Joose.A.eachR iterator
    
    [API CHANGES]

    - none
    
    [BUG FIXES]
    
    - none


3.016     2010-12-14 15:33

    [ENHANCEMENTS]
    
    - throw 'Error' instances instead of plain strings in various places
    - initial progress on reflection, allowing to receive the currently running method
    
    [API CHANGES]

    - setters no longer are being used for initial instance initialization (prior 'initialize' method)
    
    [BUG FIXES]
    
    - fixed a very specific bug, related to subclassing non-joose classes with custom constructor


3.015     2010-11-15 13:31


    [ENHANCEMENTS]
    
    - links to OpenJSAN in the documentation replaced with links to GitHub
    
    [API CHANGES]

    -none-
    
    [BUG FIXES]
    
    -none-


3.014     2010-11-01 10:32

    
    [ENHANCEMENTS]
    
    - switching to `npm` only as the distribution channel (migrating from JSAN)
    
    [API CHANGES]

    -none-
    
    [BUG FIXES]
    
    - fixed bug, when instantiating a class with traits and which is a subclass of non-joose class
      it could have been instantiated with wrong parameters

3.013     2010-10-11 15:57

    [ENHANCEMENTS]
    
    -none-
    
    [API CHANGES]

    - returned 'get/setRawValue' methods of the attribute with a bit different semantic (low-level, not documented)
    
    [BUG FIXES]
    
    - none 

3.012     2010-10-05 11:33

    [ENHANCEMENTS]
    
    - added 'hasValue/clearValue' methods for the attribute
    - removed 'get/setRawValue' methods of the attribute
    
    - minor fixes in package.json for `npm`
    
    [API CHANGES]

    -none-

    
    [BUG FIXES]
    
    -none-

3.011     2010-08-27 21:11

    [ENHANCEMENTS]
    
    - references to global RegExp object replaced with references to constructor of regexp instance,
      to allow the advanced optimizations with Google Closure optimizer (thanks to: sirrobert)
      
    - now its possible to inherit from non-Joose superclasses 
        (see http://openjsan.org/go/?l=Joose.Manual.Classes for details)
    
    - starting from this release, Joose will be also distributed via the `npm` package manager,
        > [sudo] npm install joose 
    
    - updated Joose.Manual.Installation entry
    
    [API CHANGES]

    -none-

    
    [BUG FIXES]
    
    - fixed bug, when calling `Module` over previously declared class didn't trigger a call to `body` 
    
    
3.010     2010-07-05 19:01

    [ENHANCEMENTS]
    
    - platform detection made exception-less
    
    - test suite is now passing on the NodeJS platform (has been adapted to recent changes Test.Run)
    
    [API CHANGES]

    -none-

    
    [BUG FIXES]
    
    - fixed bug, when aliases for static methods could be messed up during class mutation

3.009     2010-06-22 19:21

    [ENHANCEMENTS]
    
    - switched to Dist::Zilla for distribution management
    
    - some speed up of the core PropertySet iterators
    
    - if any value gets returned from `initialize` method, it gets returned as the result of the constructor,
      otherwise the instance itself is returned
      
    - added 'beforeConsumedBy' hook, allowing the property sets to act somehow before being consumed
    
    [API CHANGES]

    -none-

    
    [BUG FIXES]
    
    -none-


3.008   Thu, May 13, 2010

    [ENHANCEMENTS]
    
    - added ability to prefix the namespace declaration with '.' which will switch the current namespace to global
    
    - added ability to specify VERSION and AUTHORITY builders (AUTHORITY is reserved for future use with Jx.N.D)

    - added NodeJS detection

    - exporting created constructors to global NodeJS scope
    
    [API CHANGES]

    -none-
    
    
    [BUG FIXES]
    
    - fixed bug, related to fact, that typeof /a/ == 'function' in WebKit, it was preventing initializing of attributes with regexp  
    


3.007   Wed, Apr 27, 2010 

    [ENHANCEMENTS]
    
    - Class/Role/Module in the global scope aliased as Joose.Class/Joose.Role/Joose.Module
    
    - Joose no longer overwrites the Class/Role/Module symbols in global scope (should allow Prototype support)
    
    - Static instance ('my') now receives a special 'HOST' configuration parameter during instantiation, containing a hosting class constructor
    
    - Static methods are now aliased directly to the host constructor, and can be used without `my`
    
    - returning 'false' from Joose.A/O.each iterators stops the iteration and can be propagated further 
    

lib/Digest/MD5.js
//copied from author's permission:
//http://extjs.com/forum/showthread.php?p=136349#post136349

Class('Digest.MD5', {
	
    my : {
    	methods : {
    		
	        md5_hex: function(s) {
	            var hexcase = 0;
	            var chrsz = 8;
	        
	            function safe_add(x, y) {
	                var lsw = (x & 0xFFFF) + (y & 0xFFFF);
	                var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
	                return (msw << 16) | (lsw & 0xFFFF);
	            }
	            function bit_rol(num, cnt) {
	                return (num << cnt) | (num >>> (32 - cnt));
	            }
	            function md5_cmn(q, a, b, x, s, t) {
	                return safe_add(bit_rol(safe_add(safe_add(a, q), safe_add(x, t)), s), b);
	            }
	            function md5_ff(a, b, c, d, x, s, t) {
	                return md5_cmn((b & c) | ((~b) & d), a, b, x, s, t);
	            }
	            function md5_gg(a, b, c, d, x, s, t) {
	                return md5_cmn((b & d) | (c & (~d)), a, b, x, s, t);
	            }
	            function md5_hh(a, b, c, d, x, s, t) {
	                return md5_cmn(b ^ c ^ d, a, b, x, s, t);
	            }
	            function md5_ii(a, b, c, d, x, s, t) {
	                return md5_cmn(c ^ (b | (~d)), a, b, x, s, t);
	            }
	        
	            function core_md5(x, len) {
	                x[len >> 5] |= 0x80 << ((len) % 32);
	                x[(((len + 64) >>> 9) << 4) + 14] = len;
	                var a = 1732584193;
	                var b = -271733879;
	                var c = -1732584194;
	                var d = 271733878;
	                for (var i = 0; i < x.length; i += 16) {
	                    var olda = a;
	                    var oldb = b;
	                    var oldc = c;
	                    var oldd = d;
	                    a = md5_ff(a, b, c, d, x[i + 0], 7, -680876936);
	                    d = md5_ff(d, a, b, c, x[i + 1], 12, -389564586);
	                    c = md5_ff(c, d, a, b, x[i + 2], 17, 606105819);
	                    b = md5_ff(b, c, d, a, x[i + 3], 22, -1044525330);
	                    a = md5_ff(a, b, c, d, x[i + 4], 7, -176418897);
	                    d = md5_ff(d, a, b, c, x[i + 5], 12, 1200080426);
	                    c = md5_ff(c, d, a, b, x[i + 6], 17, -1473231341);
	                    b = md5_ff(b, c, d, a, x[i + 7], 22, -45705983);
	                    a = md5_ff(a, b, c, d, x[i + 8], 7, 1770035416);
	                    d = md5_ff(d, a, b, c, x[i + 9], 12, -1958414417);
	                    c = md5_ff(c, d, a, b, x[i + 10], 17, -42063);
	                    b = md5_ff(b, c, d, a, x[i + 11], 22, -1990404162);
	                    a = md5_ff(a, b, c, d, x[i + 12], 7, 1804603682);
	                    d = md5_ff(d, a, b, c, x[i + 13], 12, -40341101);
	                    c = md5_ff(c, d, a, b, x[i + 14], 17, -1502002290);
	                    b = md5_ff(b, c, d, a, x[i + 15], 22, 1236535329);
	                    a = md5_gg(a, b, c, d, x[i + 1], 5, -165796510);
	                    d = md5_gg(d, a, b, c, x[i + 6], 9, -1069501632);
	                    c = md5_gg(c, d, a, b, x[i + 11], 14, 643717713);
	                    b = md5_gg(b, c, d, a, x[i + 0], 20, -373897302);
	                    a = md5_gg(a, b, c, d, x[i + 5], 5, -701558691);
	                    d = md5_gg(d, a, b, c, x[i + 10], 9, 38016083);
	                    c = md5_gg(c, d, a, b, x[i + 15], 14, -660478335);
	                    b = md5_gg(b, c, d, a, x[i + 4], 20, -405537848);
	                    a = md5_gg(a, b, c, d, x[i + 9], 5, 568446438);
	                    d = md5_gg(d, a, b, c, x[i + 14], 9, -1019803690);
	                    c = md5_gg(c, d, a, b, x[i + 3], 14, -187363961);
	                    b = md5_gg(b, c, d, a, x[i + 8], 20, 1163531501);
	                    a = md5_gg(a, b, c, d, x[i + 13], 5, -1444681467);
	                    d = md5_gg(d, a, b, c, x[i + 2], 9, -51403784);
	                    c = md5_gg(c, d, a, b, x[i + 7], 14, 1735328473);
	                    b = md5_gg(b, c, d, a, x[i + 12], 20, -1926607734);
	                    a = md5_hh(a, b, c, d, x[i + 5], 4, -378558);
	                    d = md5_hh(d, a, b, c, x[i + 8], 11, -2022574463);
	                    c = md5_hh(c, d, a, b, x[i + 11], 16, 1839030562);
	                    b = md5_hh(b, c, d, a, x[i + 14], 23, -35309556);
	                    a = md5_hh(a, b, c, d, x[i + 1], 4, -1530992060);
	                    d = md5_hh(d, a, b, c, x[i + 4], 11, 1272893353);
	                    c = md5_hh(c, d, a, b, x[i + 7], 16, -155497632);
	                    b = md5_hh(b, c, d, a, x[i + 10], 23, -1094730640);
	                    a = md5_hh(a, b, c, d, x[i + 13], 4, 681279174);
	                    d = md5_hh(d, a, b, c, x[i + 0], 11, -358537222);
	                    c = md5_hh(c, d, a, b, x[i + 3], 16, -722521979);
	                    b = md5_hh(b, c, d, a, x[i + 6], 23, 76029189);
	                    a = md5_hh(a, b, c, d, x[i + 9], 4, -640364487);
	                    d = md5_hh(d, a, b, c, x[i + 12], 11, -421815835);
	                    c = md5_hh(c, d, a, b, x[i + 15], 16, 530742520);
	                    b = md5_hh(b, c, d, a, x[i + 2], 23, -995338651);
	                    a = md5_ii(a, b, c, d, x[i + 0], 6, -198630844);
	                    d = md5_ii(d, a, b, c, x[i + 7], 10, 1126891415);
	                    c = md5_ii(c, d, a, b, x[i + 14], 15, -1416354905);
	                    b = md5_ii(b, c, d, a, x[i + 5], 21, -57434055);
	                    a = md5_ii(a, b, c, d, x[i + 12], 6, 1700485571);
	                    d = md5_ii(d, a, b, c, x[i + 3], 10, -1894986606);
	                    c = md5_ii(c, d, a, b, x[i + 10], 15, -1051523);
	                    b = md5_ii(b, c, d, a, x[i + 1], 21, -2054922799);
	                    a = md5_ii(a, b, c, d, x[i + 8], 6, 1873313359);
	                    d = md5_ii(d, a, b, c, x[i + 15], 10, -30611744);
	                    c = md5_ii(c, d, a, b, x[i + 6], 15, -1560198380);
	                    b = md5_ii(b, c, d, a, x[i + 13], 21, 1309151649);
	                    a = md5_ii(a, b, c, d, x[i + 4], 6, -145523070);
	                    d = md5_ii(d, a, b, c, x[i + 11], 10, -1120210379);
	                    c = md5_ii(c, d, a, b, x[i + 2], 15, 718787259);
	                    b = md5_ii(b, c, d, a, x[i + 9], 21, -343485551);
	                    a = safe_add(a, olda);
	                    b = safe_add(b, oldb);
	                    c = safe_add(c, oldc);
	                    d = safe_add(d, oldd);
	                }
	                return [a, b, c, d];
	            }
	            function str2binl(str) {
	                var bin = [];
	                var mask = (1 << chrsz) - 1;
	                for (var i = 0; i < str.length * chrsz; i += chrsz) {
	                    bin[i >> 5] |= (str.charCodeAt(i / chrsz) & mask) << (i % 32);
	                }
	                return bin;
	            }
	            function binl2hex(binarray) {
	                var hex_tab = hexcase ? "0123456789ABCDEF": "0123456789abcdef";
	                var str = "";
	                for (var i = 0; i < binarray.length * 4; i++) {
	                    str += hex_tab
	                            .charAt((binarray[i >> 2] >> ((i % 4) * 8 + 4)) & 0xF)
	                            + hex_tab.charAt((binarray[i >> 2] >> ((i % 4) * 8)) & 0xF);
	                }
	                return str;
	            }
	            return binl2hex(core_md5(str2binl(s), s.length * chrsz));
	        }        
    		
    		
    	}
    }

});